Allow overlapping Zones if binding addresses are different

fturib opened the pull request:
<!--
Thank you for contributing to CoreDNS!

Please provide the following information to help us make the most of your pull request:
-->

## 1. What does this pull request do?
Change the validation of overlapping zones of  ServerBlocs, by integrating the IP address that ServerBloc is bound too. 
(so far validation of overlapping was on Zone and Port only).

The support of overlapping zones implies changes in 2 areas:

### Validation of overlapping
The validator itself is moved a bit later in the process, after BIND defined the addresses of each ServerBloc. It is now moved in the MakeServer() - after the directives are processed, before we are grouping the Configs in order to create the Servers.

NOTE: the validation is now a little more complex as it has to take care of bound zones and unbound zones. I created an object to take care of it. see zoneAddrOverlapValidator in addresse.go

### registering the Configs in dnsContext
The key used to save Config in the dnsContext has to change because the keys of ServerBloc are no more discriminant for the ServerBlocs. It was the 'zone:port' information, it is now the concat of 2 index: position of the ServerBloc in the list, position of the Key in that ServerBloc.
=> GetConfig and SaveConfig has been changed accordingly.


Unit Tests are provided for each modified areas.
I prepare integration tests to validate the workflow of DNS messages inside the BlocLists. But it involves too much changes to integrate in this PR.
Will open a PR after this one will be approved.

## 2. Which issues (if any) are related?
#1478  and #1208 
it is second part of PR #1512

## 3. Which documentation changes (if any) need to be made?
- the global README of coredns does not specify the validation of overlapping zones/ports
- do we want to add an information in that doc ?

NOTE: I did not browse yet the blogs to check if there is a mention of overlapping check.

fturib commented on core/dnsserver/register.go:
```diff
@@ -45,38 +45,41 @@ type dnsContext struct {
-			}
-			dups[za.String()] = za.String()
-
+			s.Keys[ik] = za.String()
```
Duplication test (overlapping checking) is removed from this function

fturib commented on core/dnsserver/register.go:
```diff
@@ -100,6 +103,13 @@ func (h *dnsContext) InspectServerBlocks(sourceFile string, serverBlocks []caddy
 
+	// Now that all Keys and Directives are parsed and initialized
+	// lets verify that there is no overlap on the zones and addresses to listen for
+	errValid := h.validateZonesAndListeningAddresses()
```
The overlapping test is done here.

fturib commented on core/dnsserver/register.go:
```diff
@@ -45,38 +45,41 @@ type dnsContext struct {
+	c, ok := h.keysToConfigs[key]
+	return c, ok
+}
+
```
The keys for saving Configs in the dnsContext are changed.

fturib commented on core/dnsserver/address.go:
```diff
@@ -94,3 +94,74 @@ const (
+		c.multicastOverlap[mkey] = key
+	}
+	c.registeredAddr[key] = key
+	return false, false, ""
```
Introducing an object to validate the overlaping of addresses.

We need to check that Zone x IP x Port is not overlapping, 
the tricky part is to verify that unicast cannot overlap with multicast, but 2 unicast can coexist.

fturib commented on core/dnsserver/server.go:
```diff
@@ -65,6 +65,10 @@ func NewServer(addr string, group []*Config) (*Server, error) {
+		if site.registry != nil {
+			// this config is already computed with the chain of plugin
+			continue
+		}
```
A ServerBlocList is not duplicated when  it contains a BIND with mutiple addresses (or several BIND that lead to several addresses). 
The SAME ServerBlocList that is used from all of the socket listeners to process the DNS query.

miekg commented on core/dnsserver/address.go:
```diff
@@ -94,3 +94,74 @@ const (
+	return &addrKey{keyParts[0], keyParts[1], keyParts[2], keyParts[3]}, nil
+}
+
+func (k *addrKey) isMulticast() bool {
```
Multicast? We want to care about this as well? Why? I thought this was going to be a purely textual change/check. We now also want to care about the type of addresses. 

NACK

at least in in this pr

fturib replied:
@miekg : what I name multicast is the usual key zone + port, when there is no address bound.

I guess we need to validate this overlapping use case:

not overlapping:
.:53{ bind ::1 ...}
.:53{ bind 127.0.0.1 ..}

overlapping
.:53{...}
.:53{ bind 127.0.0.1 ..}

I call ".:53" a "multicast" addrKey, but I can name it "unbound" .. 

What do you think.

NOTE: I do not introduce any new concept .. I only try to check the overlapping with existing concepts : having no bind directive, or having bind directive with one or several unicast addresses.

fturib commented:
This PR is now complete.
@miekg :let me know if my answer to your concern about "multicast" is good enough. I have renamed the variables replacing 'multicast' by 'unbound'.

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] WIP - Allow o..." ]
>This PR is now complete.
>@miekg :let me know if my answer to your concern about "multicast" is good enough. I have renamed the variables replacing 'multicast' by 'unbound'.

Don't call it multicast; that is networking term. `unbound` might be better; 
have to look at the code to make sure.

fturib commented:
It is now udpated in this PR.
There is no more reference to multicast or unicast. (I verified accross all code of coredns).
naming is now relative to 'bound' Zone/Port or 'unbound' Zone/Port

miekg commented on core/dnsserver/address.go:
```diff
@@ -94,3 +94,82 @@ const (
+
+const keyPartsSeparator = "##"
+
+type addrKey struct {
```
Why do we need this and can't extend zoneAddr (which has almost the same bits?)

fturib replied:
That was my first idea .. but ZoneAddr is used to compute/populate the Keys of ServerBloc.
and we did not wanted Address there.

I preferred to make a separated small object used only for the overlapping check. and has not the constrainst of the Keys.

miekg commented on core/dnsserver/address.go:
```diff
@@ -94,3 +94,82 @@ const (
+	}
+	mkey := k.copyAsUnbound().asKey()
+	if already, ok := c.unboundOverlap[mkey]; ok {
+		if k.isUnbound() {
```
Not completely seeing the value in having this function on only using it once(?). k.Address == "" is just a clear.

fturib replied:
it gives a name to what it is ... all is about bound and unbound in this checker

fturib replied:
but I can remove and use directly the test, with a comment

miekg commented on core/dnsserver/address.go:
```diff
@@ -94,3 +94,82 @@ const (
+		return true, false, ""
+	}
+	mkey := k.copyAsUnbound().asKey()
+	if already, ok := c.unboundOverlap[mkey]; ok {
```
as the plugin is called `bind` this is actually `bindOverlap`?

fturib replied:
I will change the name.

miekg commented:
I can't really see yet why this needs an additional 261 lines (but I haven't code it myself - so it may just balloon when you do). Anyways:

1. 

> Validation of overlapping
The validator itself is moved a bit later in the process, after BIND defined the addresses of each ServerBloc. It is now moved in the MakeServer() - after the directives are processed, before we are grouping the Configs in order to create the Servers.

This is most certainly the correct thing to do
 
2.

> NOTE: the validation is now a little more complex as it has to take care of bound zones and unbound zones. I created an object to take care of it. see zoneAddrOverlapValidator in addresse.go

Why is the key in the configs still a string? Seems to me we should just make this a struct and add the extra bind info to it? 

3.
> registering the Configs in dnsContext
The key used to save Config in the dnsContext has to change because the keys of ServerBloc are no more discriminant for the ServerBlocs. It was the 'zone:port' information, it is now the concat of 2 index: position of the ServerBloc in the list, position of the Key in that ServerBloc.
=> GetConfig and SaveConfig has been changed accordingly.

This doesn't set well with me, why do we care about the position? To make it unique, because we still store it in a map? Why can't we put it in a list then?

I think we should concentrate on step 1 first (which probably mean doing .3 as well, maybe).

fturib commented on core/dnsserver/address.go:
```diff
@@ -94,3 +94,82 @@ const (
+}
+
+func parseAddrKey(key string) (*addrKey, error) {
+	keyParts := strings.Split(key, keyPartsSeparator)
```
That is a good point. 
I do not understand why we accept all chars in Zone (when parsing the keys for instance), but the domain specified by RFC is limited to alpha + digit + hyphen.
What is the use case of # in a zone ?

fturib replied:
What is the syntax of the zone ?
I have this RFC : https://tools.ietf.org/html/rfc1035 - 2.3.1. Preferred name syntax

fturib replied:
I am interested by the reply.
But anyway, I think I can avoid to write and parse a key.

fturib commented:
About 3 - I may misunderstand your point.
FWIU, the access to the Config is through that function: `func GetConfig(c *caddy.Controller) *Config` and Controller is a caddy object (we cannot change).

if we put the Configs into a list, how to we find the right Config to return ? it has to be based on key values of the Controller, no ?

fturib commented:
Uploaded a new version:
- fix or address all comments of review (in fact point 1.)
- for point 3. - I moved the key building near to Config, but I  do not understand how we could avoid to save these configs using index here. 
NOTE: the keyToConfigs map is useful ONLY during the initialization of CoreDNS.  I cleaned it after that initialization.

about point 2 : "Why is the key in the configs still a string? Seems to me we should just make this a struct and add the extra bind info to it?" 
=> could you rephrase the comment or explain what key of what configs ... ?

miekg commented on core/dnsserver/address.go:
```diff
@@ -94,3 +101,43 @@ const (
+	unboundOverlap map[zoneAddr]zoneAddr // the unbound equiv ZoneAdddr is registered by its original key (bound or unbound)
+}
+
+func newZoneAddrOverlapValidator() *zoneAddrOverlapValidator {
```
`zoneAddrOverlapValidator` -> `zoneOverlap`

fturib replied:
ok.

miekg commented on core/dnsserver/address.go:
```diff
@@ -94,3 +101,43 @@ const (
 	TransportGRPC = "grpc"
 )
+
+// Build a Validator that rise error if the bound addresses for listeners are overlapping
```
This comment doesn't really describe what this struct does, but what you're planning to do with it, so this should be mostly moved to `registerAndCheck`, there are also a bunch of typos in it.

fturib replied:
ok. move and fix typos

miekg commented on core/dnsserver/address.go:
```diff
@@ -94,3 +101,43 @@ const (
+	return &zoneAddrOverlapValidator{registeredAddr: make(map[zoneAddr]zoneAddr), unboundOverlap: make(map[zoneAddr]zoneAddr)}
+}
+
+func (c *zoneAddrOverlapValidator) registerAndCheck(k *zoneAddr) (same bool, overlap bool, overlapKey *zoneAddr) {
```
does zoneAddr need to be pointers here?

fturib replied:
no. I change.

miekg commented on core/dnsserver/address.go:
```diff
@@ -94,3 +101,43 @@ const (
+		c.unboundOverlap[uk] = *k
+	}
+	// anyway add the current zoneAddr for future tests
+	c.registeredAddr[*k] = *k
```
when is this code exectuted?

fturib replied:
when there is no overlap. I need to keep that code.

miekg commented on core/dnsserver/config.go:
```diff
@@ -68,16 +69,22 @@ func (c *Config) HostAddresses() string {
 	return all
 }
 
+// build a key for identifying the configs during setup time
```
This isn't a proper Go comments, start iwth `keyForConfig ...`

fturib replied:
ok. update

miekg commented on core/dnsserver/register_test.go:
```diff
@@ -51,15 +51,15 @@ func TestGroupingServers(t *testing.T) {
 			failing:        false},
 
-		// 2 configs on same port, same broadcast address, diff zones -> 1 group
+		// 2 configs on same port, both undound, diff zones -> 1 group
```
- > not using bind

fturib replied:
ok. I change.

miekg commented on core/dnsserver/register_test.go:
```diff
@@ -51,15 +51,15 @@ func TestGroupingServers(t *testing.T) {
 			failing:        false},
 
-		// 2 configs on same port, same address, diff zones -> 1 group
+		// 2 configs on same port, one addressed - one unbound, diff zones -> 1 group
```
'one using bind`

fturib replied:
ok. I change.

fturib commented:
This PR has been updated, addressing the comments of review. (I hope ..).
Ready for another review...
Thanks!!

miekg commented on core/dnsserver/address.go:
```diff
@@ -94,3 +101,37 @@ const (
+	unboundOverlap map[zoneAddr]zoneAddr // the "no bind" equiv ZoneAdddr is registered by its original key
+}
+
+func newZoneValidator() *zoneOverlap {
```
newZoneOverlap

fturib replied:
oops. my bad.

miekg commented on core/dnsserver/address.go:
```diff
@@ -94,3 +101,37 @@ const (
+	return &zoneOverlap{registeredAddr: make(map[zoneAddr]zoneAddr), unboundOverlap: make(map[zoneAddr]zoneAddr)}
+}
+
+// register a new zoneAddr, return information about existing or overlapping with already registered
```
should start with registerAndCheck

fturib replied:
sure.

fturib commented:
Thanks for review and merge!

miekg commented:
Np. Thanks for your perseverance.

On 23 Feb 2018 17:21, "Francois Tur" <notifications@github.com> wrote:

> Thanks for review and merge!
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/coredns/coredns/pull/1530#issuecomment-368077202>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAVkW87B6roYMgam8SZJMlYJqgm3sAWEks5tXvOMgaJpZM4SGCFz>
> .
>
